### PR TITLE
Add scale from and to 0 support for autoscaler

### PR DESCRIPTION
**Pull request by Karthik-K-N**
**What this PR does / why we need it**:

This change adds support for scaling from zero as defined in [the enhancement](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md).
Added capacity field to status field of IBMPowerVSMachineTemplate and a separate IBMPowerVSMachineTemplate controller to update the capacity

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/833

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

- [x] Complete integration testing
- [x] Add Unit tests
- [x] Update doc

### CODE REVIEW

**Comment by Karthik-K-N** [comment]
@Prajyot-Parab @Amulyam24 @mkumatag Please take a look and help in reviewing this.

**Comment by mkumatag** [comment]
@Karthik-K-N can you please fix the CI failures?

File: api/v1beta1/suite_test.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (mkumatag) [inline api/v1beta1/suite_test.go:8]: why this change?

Comment (mkumatag) [reply api/v1beta1/suite_test.go:8]: and all the remaining copyright header change?

File: controllers/ibmpowervsmachinetemplate_controller.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	Scheme *runtime.Scheme
+}
+
+func (r *IBMPowerVSMachineTemplateReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1beta1.IBMPowerVSMachineTemplate{}).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=ibmpowervsmachinetemplates,verbs=get;list;watch;create;update;patch;delete
```
Comment (mkumatag) [inline controllers/ibmpowervsmachinetemplate_controller.go:50]: we are not creating the `ibmpowervsmachinetemplates` resource here in this `Reconcile` so are these `create;update;delete` permissions required?

File: controllers/ibmpowervsmachinetemplate_controller.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		}
+	}
+	logger.V(3).Info("machine template status", "status", machineTemplate.Status.Capacity)
+	return ctrl.Result{}, nil
+}
+
+func getIBMPowerVSMachineCapacity(machineTemplate infrav1beta1.IBMPowerVSMachineTemplate) corev1.ResourceList {
+	capacity := make(corev1.ResourceList)
+	capacity[corev1.ResourceMemory] = resource.MustParse(machineTemplate.Spec.Template.Spec.Memory)
+	capacity[corev1.ResourceCPU] = resource.MustParse(machineTemplate.Spec.Template.Spec.Processors)
```
Comment (mkumatag) [inline controllers/ibmpowervsmachinetemplate_controller.go:85]: Usually in powervs Processors value is mapped to multiples of SMT to the actual vcpus in the linux vm, hence need that logic here while calculating the capacity!

**Review by mkumatag — CHANGES_REQUESTED** [review]
fix the import formatting in the sequence mentioned here
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/774795c58fb3a730229f0773c0a8a0f77fe64f31/.golangci.yml#L58:L68

File: controllers/ibmpowervsmachinetemplate_controller.go (hunk 2 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervsmachinetemplate_controller.go]: Removed the unnecessary permission

File: controllers/ibmpowervsmachinetemplate_controller.go (hunk 3 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervsmachinetemplate_controller.go]: updated the  logical cpus

**Review by mkumatag — CHANGES_REQUESTED** [review]
overall lgtm, fix the CI

File: controllers/ibmpowervsmachinetemplate_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+		expectError            bool
+		powerVSMachineTemplate *infrav1beta1.IBMPowerVSMachineTemplate
+		expectedCapacity       corev1.ResourceList
+	}{
+		{
+			name:        "Should Reconcile successfully if no IBMPowerVSMachineTemplate found",
+			expectError: false,
+		},
+		{
+			name:                   "Should Reconcile with memory and fractional processor values",
```
Comment (Karthik-K-N) [inline controllers/ibmpowervsmachinetemplate_controller_test.go:50]: Not sure whats wrong here these two tests cases are randomly flaky. Any idea?

**Comment by Karthik-K-N** [comment]
Test is flaky https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/860/files#r972616812
/hold

File: controllers/ibmpowervsmachinetemplate_controller_test.go (hunk 4 above)
Comment (mkumatag) [reply controllers/ibmpowervsmachinetemplate_controller_test.go:50]: I could able to reproduce this issue in my local machine, I believe this is happening because patching is taking some time to reflect the actual status. This can be avoided by using `Eventually` block to test the status.

**Review by mkumatag — APPROVED** [review]
The recent changeset is working fine in my local machine as well

/lgtm
/approve

@Karthik-K-N feel free to remove hold whenever you are comfortable.

**Comment by Karthik-K-N** [comment]
Flaky tests are fixed now
/hold cancel
